ASoC: rt-sdw-common: Common functions for Realtek soundwire driver

jack-cy-yu opened the pull request:
This is the first version of common functions for Realtek soundwire codec driver.

bardliao commented on sound/soc/codecs/rt-sdw-common.c:
```diff
@@ -0,0 +1,232 @@
+{
+	if ((*buffer & 0xf0) == 0x10 || (*buffer & 0x0f) == 0x01 ||
+		(*(buffer + 1) == 0x01) || (*(buffer + 1) == 0x10))
+		return SND_JACK_BTN_2;
```
Should we pass the mask and the exact buffer to the function to get the button type accurately?

jack-cy-yu replied:
Do you mean like below style?

int rt_sdca_btn_type(unsigned char *buffer, unsigned int upper_mask, unsigned int lower_mask)
{
	if ((*buffer & upper_mask) == 0x10 || (*buffer & lower_mask) == 0x01 ||
		(*(buffer + 1) == 0x01) || (*(buffer + 1) == 0x10))
		return XXXXX;
     ...
}

bardliao replied:
Sorry, I thought the mask like 0xf0, 0x0f are codec dependent. I just realize they are the same for different codecs after reading rt712-sdca.c and rt722-sdca.c. Can multiple button types be detected? Something like `btn_type =  SND_JACK_BTN_2 | SND_JACK_BTN_0`? If so, we can't just return one button type.
My proposal is:
```c
int rt_sdca_btn_type(unsigned char *buffer) {
        u8 btn_type = 0;
        int ret = 0;

        btn_type |= buffer[0] & 0xf;
        btn_type |= (buffer[0] >> 4) & 0xf;
        btn_type |= buffer[1] & 0xf;
        btn_type |= (buffer[1] >> 4) & 0xf;

        if (btn_type & BIT(0))
                ret |= SND_JACK_BTN_2;
        if (btn_type & BIT(1))
                ret |= SND_JACK_BTN_3;
        if (btn_type & BIT(2))
                ret |= SND_JACK_BTN_0;
        if (btn_type & BIT(3))
                ret |= SND_JACK_BTN_1;

        return ret;
}
```

jack-cy-yu replied:
Got it, thanks.

bardliao commented on sound/soc/codecs/rt-sdw-common.c:
```diff
@@ -0,0 +1,232 @@
+	}
+
+	if (buf[0] == 0x11)
+		btn_type = rt_sdca_btn_type(&buf[1]);
```
It is worth to have a comment here.

jack-cy-yu replied:
Sure

Commit:
ASoC: rt-sdw-common: Common functions for Realtek soundwire driver

This is the first version of common functions for Realtek
soundwire codec driver.

Signed-off-by: Jack Yu <jack.yu@realtek.com>

lgirdwood commented on sound/soc/codecs/rt-sdw-common.c:
```diff
@@ -0,0 +1,241 @@
+int rt_sdca_index_write(struct regmap *map, unsigned int nid,
+	unsigned int reg, unsigned int value)
+{
+	unsigned int addr = (nid << 20) | reg;
+	int ret;
+
+	ret = regmap_write(map, addr, value);
+	if (ret < 0)
+		pr_err("Failed to set value: %06x <= %04x ret=%d\n",
+			addr, value, ret);
+
+	return ret;
+}
+EXPORT_SYMBOL_GPL(rt_sdca_index_write);
+
+/**
+ * rt_sdca_index_read - Read value from Realtek defined register.
+ *
+ * @map: map for setting.
+ * @nid: Realtek-defined ID.
+ * @reg: register.
+ * @value: value.
+ *
+ * A value of zero will be returned on success, a negative errno will
+ * be returned in error cases.
+ */
+int rt_sdca_index_read(struct regmap *map, unsigned int nid,
+	unsigned int reg, unsigned int *value)
+{
+	unsigned int addr = (nid << 20) | reg;
+	int ret;
+
+	ret = regmap_read(map, addr, value);
+	if (ret < 0)
+		pr_err("Failed to get value: %06x => %04x ret=%d\n",
+			addr, *value, ret);
+
+	return ret;
+}
+EXPORT_SYMBOL_GPL(rt_sdca_index_read);
```
These look generic in nature, could these be part of a generic code block ?

bardliao replied:
As far as I know, the index read/write is Realtek specific. Not sure if Cirrus use the mechanism. However, the jack type and button type detection might be generic. What do you think @shumingfan @charleskeepax ?

shumingfan replied:
@bardliao The jack type might be generic, but I'm not sure.
The button type detection will differ because the HID descriptor report might vary.

lgirdwood replied:
I see this as a low level API where the Nid/hid descriptor is just passed through to regmap, I'm good with this today though as we can continue over time to refactor as we add more SDCA logic.

charleskeepax replied:
I would be inclined to say even for the jack type stuff best to keep it in this driver for now and we can refactor as we work through the SDCA stuff.

lgirdwood reviewed (commented):
Some of the code looks very generic and could be part of a global SDCA core ?

bardliao commented:
Close it as @jack-cy-yu has submitted to upstream.
